lib/charms/loki_k8s/v0/charm_logging.py

```diff
@@ -0,0 +1,327 @@
+        server_cert = server_cert_getter(self)
+
+    if server_cert is None:
+        logger.warning(
+            f"{charm}.{server_cert_getter} returned None; sending logs over INSECURE connection."
+        )
+        return None
+
+    if not Path(server_cert).is_absolute():
+        raise ValueError(
+            f"{charm}.{server_cert_getter} should return a valid tls cert absolute path (string | Path)); "
+            f"got {server_cert} instead."
+        )
+
+    if not Path(server_cert).exists():
+        logger.warning(f"cert not found at {server_cert}: sending logs over INSECURE connection")
+        return None
+
+    return server_cert
+
+
+def _setup_root_logger_initializer(
+    charm: Type[CharmBase],
+    logging_endpoints_getter: _GetterType,
+    server_cert_getter: Optional[_GetterType],
+    service_name: Optional[str] = None,
+):
+    """Patch the charm's initializer and inject a call to set up root logging."""
+    original_init = charm.__init__
+
+    @functools.wraps(original_init)
+    def wrap_init(self: CharmBase, framework: Framework, *args, **kwargs):
+        original_init(self, framework, *args, **kwargs)
+
+        if not is_enabled():
+            logger.info("Charm logging DISABLED by env: skipping root logger initialization")
+            return
+
+        try:
+            logging_endpoints = _get_logging_endpoints(logging_endpoints_getter, self, charm)
+        except Exception:
+            # if anything goes wrong with retrieving the endpoint, we go on with logging disabled.
+            # better than breaking the charm.
+            logger.exception(
+                f"exception retrieving the logging "
+                f"endpoint from {charm}.{logging_endpoints_getter}; "
+                f"proceeding with charm_logging DISABLED. "
+            )
+            return
+
+        if not logging_endpoints:
+            return
+
+        juju_topology = JujuTopology.from_charm(self)
+        tags = {
+            **juju_topology.as_dict(),
+            "service_name": service_name or self.app.name,
+            "charm_type_name": type(self).__name__,
+            "dispatch_path": os.getenv("JUJU_DISPATCH_PATH", ""),
+        }
+        server_cert: Optional[Union[str, Path]] = (
+            _get_server_cert(server_cert_getter, self, charm) if server_cert_getter else None
+        )
+
+        root_logger = logging.getLogger()
+
+        for url in logging_endpoints:
+            handler = LokiHandler(
+                url=url,
+                tags=tags,
+                cert=str(server_cert) if server_cert else None,
+            )
+
+            root_logger.addHandler(handler)
+            root_logger.debug(
+                "Initialized charm logger",
+                extra={"tags": {"endpoint": url}},
+            )
+        return
+
+    charm.__init__ = wrap_init
+
+
+def log_charm(
+    logging_endpoints: str,
+    server_cert: Optional[str] = None,
+    service_name: Optional[str] = None,
```

Comment:
it's not the method itself,  it's the name of a method (so a string).
can't use the method itself because we're forward-referencing it (it's a decorator).